onmt/utils/multi_utils.py

```diff
@@ -0,0 +1,136 @@
+""" Pytorch Distributed utils
+    This piece of code was heavily inspired by the equivalent of Fairseq-py
+    https://github.com/pytorch/fairseq
+"""
+
+
+from __future__ import print_function
+
+import math
+import pickle
+import torch.distributed
+
+
+def is_master(opt):
+    return opt.gpu_rank == 0
+
+
+def multi_init(opt):
+    if len(opt.gpuid) == 1:
+        raise ValueError('Cannot initialize multiprocess with one gpu only')
+    dist_init_method = 'tcp://localhost:10000'
```

Comment:
Currently the dist_init_method is hardcoded. This is going to cause problems when running two different processes with multi-gpus. For example, if we have 4 gpu and we start 2 processes with 2 gpus each, then pytorch will throw the following error.

```

Traceback (most recent call last):
  File "/storage/home/ec2-user/onmt/train_multi.py", line 42, in run
    opt.gpu_rank = onmt.utils.multi_utils.multi_init(opt)
  File "/storage/home/ec2-user/onmt/onmt/utils/multi_utils.py", line 25, in multi_init
    world_size=dist_world_size, rank=opt.gpu_rank)
  File "/home/ec2-user/anaconda3/envs/pytorch_p36/lib/python3.6/site-packages/torch/distributed/__init__.py", line 94\
, in init_process_group
    group_name, rank)
RuntimeError: Address already in use at /opt/conda/conda-bld/pytorch_1524586445097/work/torch/lib/THD/process_group/G\
eneral.cpp:17
```

Reply:
I tried a simple while loop but that seemed to hang: 
``` python
21    port=10000                                                        
22    error=True                                                         
23    while error:                                                       
24        try:                                                           
25            dist_init_method = f'tcp://localhost:{port}'               
26            dist_world_size = len(opt.gpuid)                           
27            torch.distributed.init_process_group(                      
28                backend=opt.gpu_backend, init_method=dist_init_method, 
29                world_size=dist_world_size, rank=opt.gpu_rank)         
30            error=False                                                
31        except RuntimeError:                                           
32            print(f'Failed with {dist_init_method}. Trying next port') 
33            port += 13                                                 
```
Probably a better approach will to check if the tcp backend port is busy before trying to use it. That will still be susceptible to race conditions but better than the current code.

Reply:
I am not sure what command line you ran to raise that error.
it's working fine for me, I borrowed the code from fairseq btw.

Reply:
The program runs fine the first time. But if you try to run two different processes simultaneously then the problem occurs.


Regarding `fairseq` they seem to chose a port randomly to avoid collision, probably because of this same issue, see

https://github.com/pytorch/fairseq/blob/8cba7b143efd5acb14d04d7c6119028bf2e6eba2/multiprocessing_train.py#L23
 
Just for argument sake I am pasting the cmd here although I don't think the cmd is at fault here.
```
$ python /home/ec2-user/onmt//train.py -save_model ../res/t2 -gpuid 0 1 \
  -data  ../res/t2 -train_steps 34180 -valid_steps 1709 \
  -save_checkpoint_steps 1709 -copy_attn \
  -global_attention mlp -word_vec_size 128 -rnn_size 512 -layers 1 \
  -encoder_type brnn -max_grad_norm 2 -dropout 0. -batch_size 64 \
  -optim adagrad -learning_rate 0.15 -adagrad_accumulator_init 0.1 \
  -reuse_copy_attn -copy_loss_by_seqlength -bridge -seed 777
```

Reply:
I was running several instances at the same time, while exploring hyper-parameter space. I fixed it as `os.environ.get('PTH_PORT', random.randint(10000, 20000))` this way someone can override this if they really need to but otherwise a random port is selected.

Reply:
@se4u now that the multi-gpu code is merged, do you have time to PR your suggestion for this several instance use case?

Reply:
I will probably be away for at least a week because of SIGIR but I could send a PR after that. Although this is really just a single line change, if we just use `random.randint(10000, 20000)` instead of fixed `10000`, so I guess I could even make an edit and send PR from github itself. If I use an environment variable then I'll have to update the documentation as well which will be more of a hassle.

Reply:
cool question which is same to me , thanks you , but the latest version has not fixed this bug